elasticsearch_acls.go

```diff
@@ -0,0 +1,250 @@
+	path := buildPath("project", project, "service", service, "elasticsearch", "acl")
+	bts, err := h.client.doGetRequest(path, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return h.response(bts)
+}
+
+// Add function adds one new ACL record to already existing Elasticsearch ACL configuration,
+// it first checks if such ES user already exists. And if yes goes through the list of
+// ACL rules for such user and check if new rules that user tries to add already exists in
+// the configuration or not. We only add new records to the Elasticsearch configuration if case
+// there is a difference between actual existing configuration and what user is willing to add.
+func (h *ElasticSearchACLsHandler) Add(project, service string, acl ElasticSearchACL) error {
+	acls, err := h.List(project, service)
+	if err != nil {
+		return err
+	}
+
+	_, err = h.Update(project, service, ElasticsearchACLRequest{
+		ElasticSearchACLConfig: acls.ElasticSearchACLConfig.Add(acl),
+	})
+	return err
+}
+
+// Delete deletes one ACL record (might contain multiple rules) from Elasticsearch ACLs list
+func (h *ElasticSearchACLsHandler) Delete(project, service string, acl ElasticSearchACL) error {
+	acls, err := h.List(project, service)
+	if err != nil {
+		return err
+	}
+
+	_, err = h.Update(project, service,
+		ElasticsearchACLRequest{
+			ElasticSearchACLConfig: acls.ElasticSearchACLConfig.Delete(acl),
+		},
+	)
+	return err
+}
+
+// response checks if response fom Aiven API contains any errors
+func (h *ElasticSearchACLsHandler) response(r []byte) (*ElasticSearchACLResponse, error) {
+	var rsp *ElasticSearchACLResponse
+	if err := json.Unmarshal(r, &rsp); err != nil {
+		return nil, err
+	}
+
+	// response cannot be empty
+	if rsp == nil {
+		return nil, ErrNoResponseData
+	}
+
+	// check API response errors
+	if rsp.Errors != nil && len(rsp.Errors) != 0 {
+		return nil, errors.New(rsp.Message)
+	}
+
+	return rsp, nil
+}
+
+// Delete subtracts ACL from already existing Elasticsearch ACLs config
+func (conf ElasticSearchACLConfig) Delete(acl ElasticSearchACL) ElasticSearchACLConfig {
+	for p, existingAcl := range conf.ACLs { // subtract ALC from existing ACLs config entry that supposed to be deleted
+		if acl.Username == existingAcl.Username {
+			for i := range existingAcl.Rules {
+				// remove ACL from existing ACLs list
+				for _, rule := range acl.Rules {
+					if existingAcl.Rules[i].Permission == rule.Permission && existingAcl.Rules[i].Index == rule.Index {
+						conf.ACLs[p].Rules = append(conf.ACLs[p].Rules[:i], conf.ACLs[p].Rules[i+1:]...)
+					}
+				}
+
+				// delete ACL item from ACLs list is there are not rules attached to it
+				if len(conf.ACLs[p].Rules) == 0 {
+					conf.ACLs = append(conf.ACLs[:p], conf.ACLs[p+1:]...)
+				}
+			}
+		}
+	}
+
+	return conf
+}
+
+// Check checks Elasticsearch permission and rules
+func (conf ElasticSearchACLConfig) Check() error {
```

Comment:
I would recommend dropping this as the API will do the checking

Reply:
done